src/python/pants/option/global_options.py

```diff
@@ -1752,6 +1752,20 @@ class GlobalOptions(BootstrapOptions, Subsystem):
         default=[],
     )
 
+    use_deprecated_lambdex_layout = BoolOption(
```

Comment:
Why is this a global option? In fact, why do we need this option at all? 

I might be missing something, but I think it would be simpler to do the relevant check in the code that uses the subsystem, and issue the deprecation warning there? 

I'm not sure why this is a complicated as it is, so maybe I'm not seeing something.

Reply:
I wanted to allow users (including my work repo) to opt in to the new `layout="zip"`-by-default behaviour early and globally. In particular, I wouldn't want to have to set it on each target when I upgrade to 2.18... and then go through and delete them all in 2.19, once the default flips. I also thought it would be nice to allow users to silence the warning, but that's less critical.

I suppose another option would be: unconditional (but memoized etc.) warning as it is, and users can add a `__defaults__` in the root `BUILD` to reduces the updates required. This does mean pants won't be able prompt users to remove it when it's not necessary, but that's probably okay, I can have the warning suggest something with a comment.

I'm more than happy to be guided to simpler options, if you think I'm solving problems that don't need to be solved!

(Side note, is 2.19 the right release to remove? Maybe it should be a longer deprecation period?)

Reply:
There is a very low chance that switching from lambdex to zip will have any negative consequence, so I think 2.19.x is fine!

But I'm sure we can find a much simpler way of deprecating. For one thing, maybe (and I wish I'd thought of this when reviewing that change, sorry) we don't need a per-target `layout` at all, and this can be a repo-wide setting? After all, we're about to get rid of it entirely, so the intermediate state of "some lambdas use lambdex and some use zip" seems unnecessary. Esp. since we expect zero user-noticeable difference except better performance (and I guess the name change of the handler).  That change has not been in any release yet, so not too late to replace it.

Instead, we can put layout on the `lambdex` scope (for want of a better one), and default it to `lambdex`. But, if you rely on that default in 2.17.x (I think a cherry-pick is fine for this) you get a warning that you should set it explicitly and than `lambdex` is going away. Then we make `zip` the default in 2.18.x, with a warning if you set to `lambdex` (or if you set explicitly to `zip`, since that is now the default, and the option is going away), and then we remove the option entirely in 2.19.x.

Does that make sense? This is often how we do stuff like this - introduce an option to deprecate a feature, then deprecation the option one release later.

Either way, this shouldn't be on the global options, because it is very specific to a backend.

Reply:
> For one thing, maybe (and I wish I'd thought of this when reviewing that change, sorry) we don't need a per-target layout at all, and this can be a repo-wide setting?
>
> Does that make sense? This is often how we do stuff like this - introduce an option to deprecate a feature, then deprecation the option one release later.

Yeah, that makes a lot of sense.

Simplifying layout to a global option is way better 👍  I had originally been thinking of other layouts that do make sense per-target (e.g. `layout="directory"` to not have it zipped, or even `layout="layer"`), but that's all a total distraction from the core of "move away from Lambdex" and should be handled independently.

Next steps for me:

1. revert #19022, because it'll make the next steps cleaner: #19071
2. set-up the global `layout = "lambdex"` setting
3. redo #19022, getting rid of the `layout` field and using the global setting instead

Thanks for taking the time to think through a better approach!